### PR TITLE
Autoincrement 2

### PR DESCRIPTION
**Pull request by wibeasley**
(Hi @krlmlr, this PR replaces #58)

Can someone confirm that this is the intended behavior?  I'd hate to be writing tests that approve an unintended effect.

When the autoincrement column isn't in the data.frame sent to the database, it's populated by SQLite.

But when it is populated before getting to the database, SQLite won't assign/update values.  And when retrieved from SQLite, the data.frame will be re-ordered (corresponding to the values in the (ignored) autoincrement column).

```r
> ds_local
    id name score
1  126    a     1
2  125    b     2
3  124    c     3
...
24 103    x    24
25 102    y    25
26 101    z    26

> dbWriteTable(con, name = 'tbl', value = ds_local, append = TRUE, row.names = FALSE)
> ds_remote <- dbReadTable(con, "tbl")

> ds_remote
    id name score
1  101    z    26
2  102    y    25
3  103    x    24
...
24 124    c     3
25 125    b     2
26 126    a     1
```

### CODE REVIEW

**Comment by krlmlr** [comment]
SQL tables don't have an intrinsic order unless you supply `ORDER BY` with the query (which `dbReadTable()` doesn't). So the SQL engine is free to do whatever it wants to in this case.

I would expect the following behavior:

- If the autoincrement column doesn't exist in the data frame passed to `dbWriteTable()`, it's populated by the SQL engine
- If it exists, the values are taken from the data; NA values should be converted to explicit NULL in this case, and it's up to the SQL engine to decide if further conversion happens there

Does this align with your observations?

**Comment by wibeasley** [comment]
@krlmlr, yes, that is consistent.  And to be specific on the second bullet, the engine apparently decides to autopopulate the column of NAs (which makes sense to me).

```r
>   ds_local <- data.frame(
+     #Notice the 'id' column is set locally, which overrides the autoincrement assignment in the DB.
+     id               = NA_integer_,
+     name             = letters,
+     score            = 1:26,
+     stringsAsFactors = FALSE
+   )

>   dbWriteTable(con, name = 'tbl', value = ds_local, append = TRUE, row.names = FALSE)
[1] TRUE
>   
>   ds_remote <- dbReadTable(con, "tbl")
>   ds_remote
   id name score
1   1    a     1
2   2    b     2
3   3    c     3
...
24 24    x    24
25 25    y    25
26 26    z    26
```

**Comment by krlmlr** [comment]
Changes minus roxygen noise: https://github.com/rstats-db/RSQLite/compare/7997aca30ba98d994571bbef065ff9253f4eefde...24d323639db901d8966ea551a17ce87c93ef6816

File: R/table.R [hunk 1]
```diff
@@ -27,6 +27,10 @@ NULL
 #' @param field.types character vector of named  SQL field types where
 #'   the names are the names of new table's columns. If missing, types inferred
 #'   with \code{\link[DBI]{dbDataType}}).
+#' @details If a column is specified \code{AUTOINCREMENT} in the database and
+#'   you wish to allow the database to automatically assign a value, the column
+#'   must exist (in the \code{data.frame} passed to the write function's 
+#'   \code{value} parameter), and be filled with \code{NA} values.
```
Comment (krlmlr) [inline R/table.R]: Does this also apply for append = TRUE?

File: tests/testthat/test-dbWriteTable.R [hunk 2]
```diff
@@ -94,3 +94,78 @@ test_that("can round-trip utf-8", {
...
+    score            = 1:26,
+    stringsAsFactors = FALSE
+  )
+  
+  dbSendQuery(con, ddl)  
+  dbWriteTable(con, name = 'tbl', value = ds_local, append = TRUE, row.names = FALSE)
+  
+  ds_remote <- dbReadTable(con, "tbl")
+  
+  expect_equal(ds_remote[, c("name", "score")], ds_local[, c("name", "score")], label = "The two non-autoincrement columns should be correct.")
```
Comment (krlmlr) [inline tests/testthat/test-dbWriteTable.R]: Use of label = is discouraged, in favor of test expressions that speak for themselves: https://github.com/hadley/testthat/issues/218.

File: tests/testthat/test-dbWriteTable.R [hunk 3]
```diff
@@ -94,3 +94,78 @@ test_that("can round-trip utf-8", {
...
+    score            = 1:26,
+    stringsAsFactors = FALSE
+  )
+  
+  dbSendQuery(con, ddl)  
+  dbWriteTable(con, name = 'tbl', value = ds_local, append = TRUE, row.names = FALSE)
+  
+  ds_remote <- dbReadTable(con, "tbl")
+  
+  expect_equal(ds_remote$name,  ds_local$name[26:1],   label = "The two non-autoincrement columns should be correct (and reversed).")
```
Comment (krlmlr) [inline tests/testthat/test-dbWriteTable.R]: I think it's more robust to order ds_result by score before comparing.

**Comment by wibeasley** [comment]
@krlmlr 
* I rewrote some of the tests so labels aren't necessary.  I kept the labels in other places where the intent is less clear without it (and I couldn't think of a way to rewrite it).
* Datasets are now sorted before the tests (in response to https://github.com/rstats-db/RSQLite/pull/119/files#r44763094)
* Two more scenarios are tested (where the `id` has some values and some `NA`s).  As before, I think everything is working as people intended.  The new tests are included more for the sake of detecting future regressions.

*Question:*
In [response to this question of yours](https://github.com/rstats-db/RSQLite/pull/119/files#r44763094), do you want the `append=TRUE` and `append=FALSE` comparisons to be within the same test/function?  In other words, should the function "autoincrement correctly populated by database" address both scenarios?  Or should they have separate set ups?  

I'd favor separation/isolation, even though it would increase the amount of code. But I wanted to check with you guys first, since you guys tend to have strong opinions about these things.

File: tests/testthat/test-dbWriteTable.R [hunk 4]
```diff
@@ -94,3 +94,142 @@ test_that("can round-trip utf-8", {
...
+  ds_remote <- ds_remote[order(ds_remote$id), ] #Sort returned data.frame so comparisons are more robust.
+  
+  expected_ids <- c(101, 102, 103, 204, 205, 306, 307) #Notice the jumps to 204 and to 306.
+  
+  expect_equal(ds_remote$name,  ds_local$name)
+  expect_equal(ds_remote$score, ds_local$score)
+  expect_equal(ds_remote$id, expected_ids, label = "The autoincrement values should be assigned correctly.")
+})
+
+test_that("autoincrement populated before database with some NAs in reverse order", {
```
Comment (krlmlr) [inline tests/testthat/test-dbWriteTable.R]: "in reverse order": There's no intrinsic order.

File: R/table.R (hunk 1 above)
Comment (krlmlr) [inline R/table.R]: You don't need to have the column in the data frame if the table already exists. Your text seems to imply otherwise.

**Comment by krlmlr** [comment]
Very thorough tests, thanks! Isolated tests are fine with me.

Could you please also add a bullet to NEWS.md?

**Comment by wibeasley** [comment]
* > ["in reverse order": There's no intrinsic order. Same above.](https://github.com/rstats-db/RSQLite/pull/119#discussion_r44797022)

    That test description was intended to communicate that the `id` assignment in the local dataset was (approximately) reverse ordered ...and thus exposed the potential for a sequential-assigning-database to duplicate previous values.  I see your point, and will try to think of a less ambiguous phrase.  In the meantime, I'm open to suggestions if one has already occurred to you.
    ```r
    c(NA_integer_, 306, NA_integer_, -204, 103, 102, NA_integer_)
    ```

* > [You don't need to have the column in the data frame if the table already exists. Your text seems to imply otherwise.](https://github.com/rstats-db/RSQLite/pull/119/files#r44797259)

    Agreed.  My comfort and understanding of this functionality is a lot different than it was [a year ago](https://github.com/rstats-db/RSQLite/pull/58) when I wrote the description.  I'll update the documentation after I write the `append=FALSE` tests.  I may learn something that informs the documentation and makes it clearer.

* > Could you please also add a bullet to NEWS.md?

    Sure, I'd forgotten about that.  Thanks for being so polite while making good suggestions.

**Comment by wibeasley** [comment]
@krlmlr, I believe I've made all the modifications you've requested.  Please tell me if there's something else I can do.

I didn't add any tests for `append=FALSE` in this PR because [it's not possible unless overwrite is true](https://github.com/rstats-db/RSQLite/blob/322a3fc14aafc2bc3db0337c0c16e46edead4a46/R/table.R#L52-L55), and I assume at autoincrement DDL is irrelevant when overwriting.  I have started a batch of overwrite tests, but I think those are different enough to justify a different PR.

**Comment by krlmlr** [comment]
Thanks, looks good to me. I think overwrite = TRUE tests are generic enough for the new DBItest package.

@hadley: Can we merge this?

**Comment by krlmlr** [comment]
Thanks!

**Pull request merged**